cpp/test/sg/fil_child_index_test.cu

```diff
@@ -0,0 +1,249 @@
+#include <ostream>
+#include <utility>
+
+namespace ML {
+
+using namespace fil;
+
+struct proto_inner_node {
+  bool def_left = false, is_categorical = false;
+  int fid      = 0;     // feature id
+  int set      = 0;     // which bit set represents the matching category list
+  float thresh = 0.0f;  // threshold
+  int left     = 1;     // left child idx
+  val_t split()
+  {
+    val_t split;
+    if (is_categorical)
+      split.idx = set;
+    else
+      split.f = thresh;
+    return split;
+  }
+  operator sparse_node16()
+  {
+    return sparse_node16({}, split(), fid, def_left, false, is_categorical, left);
+  }
+  operator sparse_node8()
+  {
+    return sparse_node8({}, split(), fid, def_left, false, is_categorical, left);
+  }
+  operator dense_node() { return dense_node({}, split(), fid, def_left, false, is_categorical); }
+};
+
+std::ostream& operator<<(std::ostream& os, const proto_inner_node& node)
+{
+  os << "def_left " << node.def_left << " is_categorical " << node.is_categorical << " fid "
+     << node.fid << " set " << node.set << " thresh " << node.thresh << " left " << node.left;
+  return os;
+}
+
+// proto inner node
+#define NODE(...)                                               \
+  []() {                                                        \
+    struct NonDefaultProtoInnerNode : public proto_inner_node { \
+      NonDefaultProtoInnerNode() { __VA_ARGS__; }               \
+    };                                                          \
+    return proto_inner_node(NonDefaultProtoInnerNode());        \
+  }()
+
+// proto category sets for one node
+struct ProtoCategorySets {
+  // each bit set for each feature id is in a separate vector
+  // read each uint8_t from right to left, and the vector(s) - from left to right
+  std::vector<std::vector<uint8_t>> bits;
+  std::vector<int> max_matching;
+  operator cat_sets_owner()
+  {
+    ASSERT(
+      bits.size() == max_matching.size(),
+      "internal error: ProtoCategorySets::bits.size() != ProtoCategorySets::max_matching.size()");
+    std::vector<uint8_t> flat;
+    for (std::vector<uint8_t> v : bits) {
+      for (uint8_t b : v)
+        flat.push_back(b);
+    }
+    return {flat, max_matching};
+  }
+};
+
+struct ChildIndexTestParams {
+  proto_inner_node node;
+  int parent_node_idx = 0;
+  cat_sets_owner cso;
+  float input = 0.0f;
+  int correct = INT_MAX;
+};
+
+std::ostream& operator<<(std::ostream& os, const ChildIndexTestParams& ps)
+{
+  os << "node = {\n"
+     << ps.node << "\n} "
+     << "parent_node_idx = " << ps.parent_node_idx << " cat_sets_owner = {\n"
+     << ps.cso << "\n} input = " << ps.input << " correct = " << ps.correct;
+  return os;
+}
+
+/** mechanism to use named aggregate initialization before C++20, and also use
+    the struct defaults. Using it directly only works if all defaulted
+    members come after ones explicitly mentioned.
+**/
+#define CHILD_INDEX_TEST_PARAMS(...)                                      \
```

Comment:
I'm not entirely sold on introducing a macro here to make use of designated initializers in tests. Can you give me some sense of why we need that feature enough to justify a macro?

Reply:
Could we not introduce a helper function that would effectively reduce boilerplate in the same way?

Reply:
added the comment re: macro to each of the two instances in this file and added you to a thread. Also copied my main arguments into the review comment re: `NODE(...)` instance